spec/consensus/reactor/reactor.md

```diff
@@ -0,0 +1,355 @@
+It is in this layer that partial synchrony assumptions are combined with P2P primitives to provide Tendermint BFT's liveness guarantees.
+Exchanges between State and Communication use multiple forms, but we will call them all **Gossip-I** here.
+
+
+```
+...
+==========ABCI=========
+
+  |     State       |
+  |.....Gossip-I....| Consensus Reactor
+  |  Communication  |
+
+- - - - - P2P-I - - - -
+         P2P
+=======================
+    Network Stack
+```
+
+The goal here is to understand what the State layer requires from the Communication layer (Gossip-I) and what the Communication layer requires from P2P (P2P-I) in order to satisfy the State layer needs.
+
+# Status
+
+This is a Work In Progress. It has not been reviewed and is far from completion. 
+
+Some descriptions only resemble TLA+. They will be rigorously rewritten later.
+
+
+
+# Outline
+> **TODO**
+
+
+
+# Part 1: Background
+
+Here we assume that you understand the Tendermint BFT algorithm, which has been described in multiple places, such as [here](../).
+
+The Tendermint BFT algorithm assumes that a **Global Stabilization Time (GST)** exists, after which communication is reliable and timely:
+
+> **Eventual $\Delta$-Timely Communication**: There is a bound $\Delta$ and an instant GST (Global Stabilization Time) such that if a correct process $p$ sends message $m$ at time $t \geq \text{GST}$ to a correct process $q$, then $q$ will receive $m$ before $t + \Delta$.
+
+Tendermint BFT also assumes that this property is used to provide **Gossip Communication**:
+
+> **Gossip communication**: (i)If a correct process $p$ sends some message $m$ at time $t$, all correct processes will receive $m$ before $\text{max} \{t,\text{GST}\} + \Delta$.
+Furthermore, (ii)if a correct process $p$ receives some message $m$ at time $t$, all correct processes will receive $m$ before $\text{max}\{t,\text{GST}\} + \Delta$.
+
+Because Gossip Communication requires even messages sent before GST to be reliably delivered between correct processes ($t$ may happen before GST in (i)) and because GST could take arbitrarily long to arrive, in practice, implementing this property would require an unbounded message buffer.
+
+However, while Gossip Communication is a sufficient condition for Tendermint BFT to terminate, it is not strictly necessary that all messages from correct processes reach all other correct processes.
+What is required is for either the message to be delivered or that, eventually, some newer message, with information that supersedes that in the first message, to be timely delivered.
+In Tendermint BFT, this property is seen, for example, when nodes ignore proposal messages from prior rounds.
+
+> **Supersession**: We say that a message $lhs$ supersedes a message $rhs$ if after receiving $lhs$ a process would make at least as much progress as it would by receiving $rhs$ and we note is as follows: $\text{lhs} \overset{s}{>} \text{rhs}$.   
```

Comment:
In the past we looked a bit into communication closure w.r.t. consensus. Roughly, the lexicographical order over the tuples (height, round, step) defines a notion of logical time, and when I am in a certain height, round and step, I don't care about messages from "the past". Tendermint consensus is mostly communication-closed in that we don't care about messages from the past. An exception is line 28 in the arXiv paper where we accept prevote messages from previous rounds vr for the same height.

I guess a precise constructive definition of "supersession" can be done along these lines.